Potentially fixing the bug of incorrect deep_free

MMZK1526 opened the pull request:
1. Store the offset between head and payload in a global variable.
2. Use this offset, not sizeof(), to get the right position of head.
3. The bug was originally reproducible on my 64bit mac, now it's fixed.
May require further testing on a 32bit machine as well.
4. Further fix a bug on the inconsistency on block_size that causes
a few bytes not to be freed in every cycle.
5. Known issue: fast_free causes segfault; fast_bins are always NULL.

chinesebear commented on include/deep_mem.h:
```diff
@@ -12,7 +12,7 @@
 #define P_FLAG_MASK (0x00000002)
 #define BLOCK_SIZE_MULTIPLIER (3)
-#define BLOCK_SIZE_MASK (0xfffffff4)
+#define BLOCK_SIZE_MASK (0xfffffffc)
```
this bit mask error is root cause?

MMZK1526 replied:
No, but I've checked with Joey; this should be fffffffc

chinesebear replied:
done!

chinesebear commented on src/deep_mem.c:
```diff
@@ -207,40 +218,38 @@ deep_malloc_fast_bins (uint32_t size)
 deep_malloc_sorted_bins (uint32_t size)
 {
-  block_size_t aligned_size = ALIGN_MEM_SIZE (size + sizeof (block_head_t));
+  block_size_t aligned_size = ALIGN_MEM_SIZE (size + 4);
```
4 is magic word. can you change with macro value?

chinesebear commented on src/deep_mem.c:
```diff
@@ -308,14 +313,17 @@ deep_free_sorted_bins (void *ptr)
+      
       _merge_into_single_block (the_other, block);
-
+puts("test");
```
puts() is necessary?

MMZK1526 replied:
This and above are used during debugging; I'll change them back and remove unnecessary prints

chinesebear reviewed (commented):
some ques

MMZK1526 committed:
Fix payload size inconsistency in malloc and free

MMZK1526 committed:
Fix deep_free_fast_bins NULL ptr error.
1. Do a NULL check after accessing nullable array element.
2. Computes free memory size correctly in deep_malloc_fast_bins.
3. Add tests for various malloc size.
